File: app/src/main/java/com/noxvision/app/hunting/weather/WeatherApiClient.kt
package com.noxvision.app.hunting.weather

import com.noxvision.app.hunting.database.entities.CachedWeather
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import com.noxvision.app.BuildConfig

class WeatherApiClient(private val apiKey: String = DEFAULT_API_KEY) {

    private val jsonParser = Json { ignoreUnknownKeys = true }

    companion object {
        // OpenWeatherMap Free Tier API Key loaded from local.properties via BuildConfig
        private val DEFAULT_API_KEY = BuildConfig.OPENWEATHER_API_KEY
        private const val BASE_URL = "https://api.openweathermap.org/data/2.5/weather"
    }

    suspend fun fetchWeather(latitude: Double, longitude: Double): Result<CachedWeather> {
        return withContext(Dispatchers.IO) {
            try {
                val url = URL("$BASE_URL?lat=$latitude&lon=$longitude&appid=$apiKey&units=metric&lang=de")
                val connection = url.openConnection() as HttpURLConnection
                connection.connectTimeout = 10000
                connection.readTimeout = 10000
                connection.requestMethod = "GET"

                val responseCode = connection.responseCode
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    val response = connection.inputStream.bufferedReader().use { it.readText() }
                    connection.disconnect()

                    val weather = parseWeatherResponse(response, latitude, longitude)
                    Result.success(weather)
                } else {
                    connection.disconnect()
                    Result.failure(Exception("HTTP Error: $responseCode"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    internal fun parseWeatherResponse(json: String, latitude: Double, longitude: Double): CachedWeather {
        val response = jsonParser.decodeFromString<WeatherResponse>(json)

        val main = response.main
        val wind = response.wind
        val clouds = response.clouds
        val sys = response.sys
        val weather = response.weather[0]

        return CachedWeather(
            id = 0,
            timestamp = System.currentTimeMillis(),
            latitude = latitude,
            longitude = longitude,
            temperature = main.temp,
            feelsLike = main.feelsLike,
            humidity = main.humidity,
            pressure = main.pressure,
            windSpeed = wind.speed,
            windDirection = wind.deg ?: 0,
            windGust = wind.gust?.takeIf { it > 0.0 },
            cloudiness = clouds.all,
            visibility = response.visibility ?: 10000,
            description = weather.description,
            icon = weather.icon,
            sunrise = sys.sunrise * 1000,
            sunset = sys.sunset * 1000
        )
    }
}

File: NoxVision-iOS/NoxVision/Hunting/Weather/WeatherApiClient.swift
import Foundation

class WeatherApiClient {
    // In production, store this in a secure config or keychain
    private let apiKey: String
    private static let baseURL = "https://api.openweathermap.org/data/2.5/weather"

    init(apiKey: String = "") {
        // Load from Info.plist or environment
        if apiKey.isEmpty {
            self.apiKey = Bundle.main.object(forInfoDictionaryKey: "OPENWEATHER_API_KEY") as? String ?? ""
        } else {
            self.apiKey = apiKey
        }
    }

    func fetchWeather(latitude: Double, longitude: Double) async -> Result<CachedWeather, Error> {
        let sanitizedKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines).replacingOccurrences(of: "\"", with: "")
        guard !sanitizedKey.isEmpty else {
            return .failure(NSError(domain: "WeatherAPI", code: -1, userInfo: [NSLocalizedDescriptionKey: "OPENWEATHER_API_KEY is missing"]))
        }

        guard let url = URL(string: "\(Self.baseURL)?lat=\(latitude)&lon=\(longitude)&appid=\(sanitizedKey)&units=metric&lang=de") else {
            return .failure(NSError(domain: "WeatherAPI", code: -2, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"]))
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 10.0

        do {
            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                return .failure(NSError(domain: "WeatherAPI", code: -3, userInfo: [NSLocalizedDescriptionKey: "Invalid response"]))
            }

            guard httpResponse.statusCode == 200 else {
                let errorBody = String(data: data, encoding: .utf8)?.prefix(200) ?? ""
                return .failure(NSError(domain: "WeatherAPI", code: httpResponse.statusCode,
                    userInfo: [NSLocalizedDescriptionKey: "HTTP Error: \(httpResponse.statusCode) (\(errorBody))"]))
            }

            let weather = try parseWeatherResponse(data, latitude: latitude, longitude: longitude)
            return .success(weather)
        } catch {
            return .failure(error)
        }
    }

    func parseWeatherResponse(_ data: Data, latitude: Double, longitude: Double) throws -> CachedWeather {
        let decoder = JSONDecoder()
        let response = try decoder.decode(WeatherResponse.self, from: data)

        let main = response.main
        let wind = response.wind
        let clouds = response.clouds
        let sys = response.sys
        let condition = response.weather.first ?? WeatherCondition(description: "", icon: "")

        return CachedWeather(
            timestamp: Date(),
            latitude: latitude,
            longitude: longitude,
            temperature: main.temp,
            feelsLike: main.feelsLike,
            humidity: main.humidity,
            pressure: main.pressure,
            windSpeed: wind.speed,
            windDirection: wind.deg ?? 0,
            windGust: wind.gust.flatMap { $0 > 0 ? $0 : nil },
            cloudiness: clouds.all,
            visibility: response.visibility ?? 10000,
            weatherDescription: condition.description,
            icon: condition.icon,
            sunrise: Date(timeIntervalSince1970: TimeInterval(sys.sunrise)),
            sunset: Date(timeIntervalSince1970: TimeInterval(sys.sunset))
        )
    }
}
